File: vox/typescript/packages/roam-postcard/src/roam_error.ts
// TODO: move to roam-wire

import { encodeVarint } from "./binary/varint.ts";

export const ROAM_ERROR = {
  USER: 0,
  UNKNOWN_METHOD: 1,
  INVALID_PAYLOAD: 2,
  CANCELLED: 3,
} as const;

export function encodeUnknownMethod(): Uint8Array {
  return encodeVarint(ROAM_ERROR.UNKNOWN_METHOD);
}

export function encodeInvalidPayload(): Uint8Array {
  return encodeVarint(ROAM_ERROR.INVALID_PAYLOAD);
}

File: vox/rust/roam-types/src/roam_error.rs
use facet::Facet;

// r[rpc.fallible.roam-error]
/// Protocol-level error wrapper distinguishing application errors from roam infrastructure errors.
///
/// On the caller side, all return types are wrapped as `Result<T, RoamError<E>>`:
///   * Infallible `fn foo() -> T` becomes `Result<T, RoamError>`
///   * Fallible `fn foo() -> Result<T, E>` becomes `Result<T, RoamError<E>>`
#[derive(Debug, Clone, Facet)]
#[repr(u8)]
pub enum RoamError<E = ::core::convert::Infallible> {
    /// The handler ran and returned an application error.
    User(E),

    /// No handler recognized the method ID.
    UnknownMethod,

    /// The arguments could not be deserialized.
    InvalidPayload(String),

    /// The call was cancelled before completion (e.g. handler dropped without replying).
    Cancelled,

    /// The underlying connection closed while the call was in flight.
    ConnectionClosed,

    /// The session shut down while the call was in flight.
    SessionShutdown,

    /// The call could not be sent because the transport is dead.
    SendFailed,

    /// The runtime refused to guess after recovery.
    Indeterminate,
}
